src/plugins/interactive_setup/server/elasticsearch_service.ts

```diff
@@ -0,0 +1,239 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import type { ApiResponse } from '@elastic/elasticsearch';
+import { errors } from '@elastic/elasticsearch';
+import type { Duration } from 'moment';
+import type { Observable } from 'rxjs';
+import { from, of, timer } from 'rxjs';
+import {
+  catchError,
+  distinctUntilChanged,
+  exhaustMap,
+  map,
+  shareReplay,
+  takeWhile,
+} from 'rxjs/operators';
+
+import type {
+  ElasticsearchClientConfig,
+  ElasticsearchServicePreboot,
+  ICustomClusterClient,
+  Logger,
+  ScopeableRequest,
+} from 'src/core/server';
+
+import { ElasticsearchConnectionStatus } from '../common';
+import { getDetailedErrorMessage } from './errors';
+
+interface EnrollParameters {
+  apiKey: string;
+  hosts: string[];
+  // TODO: Integrate fingerprint check as soon core supports this new option:
+  // https://github.com/elastic/kibana/pull/108514
+  caFingerprint?: string;
+}
+
+export interface ElasticsearchServiceSetupDeps {
+  /**
+   * Core Elasticsearch service preboot contract;
+   */
+  elasticsearch: ElasticsearchServicePreboot;
+
+  /**
+   * Interval for the Elasticsearch connection check (whether it's configured or not).
+   */
+  connectionCheckInterval: Duration;
+}
+
+export interface ElasticsearchServiceSetup {
+  /**
+   * Observable that yields the last result of the Elasticsearch connection status check.
+   */
+  connectionStatus$: Observable<ElasticsearchConnectionStatus>;
+
+  /**
+   * Iterates through provided {@param hosts} one by one trying to call Kibana enrollment API using
+   * the specified {@param apiKey}.
+   * @param apiKey The ApiKey to use to authenticate Kibana enrollment request.
+   * @param hosts The list of Elasticsearch node addresses to enroll with. The addresses are supposed
+   * to point to exactly same Elasticsearch node, potentially available via different network interfaces.
+   */
+  enroll: (params: EnrollParameters) => Promise<EnrollResult>;
+}
+
+/**
+ * Result of the enrollment request.
+ */
+export interface EnrollResult {
+  /**
+   * Host address of the Elasticsearch node that successfully processed enrollment request.
+   */
+  host: string;
+  /**
+   * PEM CA certificate for the Elasticsearch HTTP certificates.
+   */
+  ca: string;
+  /**
+   * Service account token for the "elastic/kibana" service account.
+   */
+  serviceAccountToken: { name: string; value: string };
+}
+
+export class ElasticsearchService {
```

Comment:
__note:__ I finally moved this to a "Service" since it has a clear lifetime with appropriate setup/stop hooks.